NPM Packages Table on Linux

Exbia opened the pull request:
Adding query support and table for NPM packages on Linux

obelisk commented:
ok to test

obelisk commented on osquery/tables/system/linux/npm_packages.cpp:
```diff
@@ -0,0 +1,59 @@
+namespace osquery {
+namespace tables {
+
+const std::map<std::string, std::string> PackageTopLevelKeys{
```
constants should start with a k, ie kPackageTopLevelKeys

obelisk commented on osquery/tables/system/linux/npm_packages.cpp:
```diff
@@ -0,0 +1,59 @@
+#include "osquery/core/conversions.h"
+
+namespace fs = boost::filesystem;
+namespace pt = boost::property_tree;
```
Please don't use ptree anymore and use the RapidJSON functions in conversions.

obelisk commented on osquery/tables/system/linux/npm_packages.cpp:
```diff
@@ -0,0 +1,59 @@
+  resolveFilePattern("/usr/lib/node_modules/%/package.json", paths);
+
+  for (const auto& path : paths) {
+    pt::ptree tree;
```
Please use JSON documents

obelisk commented on osquery/tables/system/linux/npm_packages.cpp:
```diff
@@ -0,0 +1,59 @@
+  QueryData results;
+
+  std::vector<std::string> paths;
+  resolveFilePattern("/usr/lib/node_modules/%/package.json", paths);
```
If you have this passed in via `WHERE` clause then we can get this info from any directory on the system or fall back to global is one isn't provided.

Exbia commented:
I've removed the use of ptree for reading the package.json file, and re-implemented it with JSON Documents. 

I've also added a 'directory' column which can be used in WHERE filters.

With these changes, we can now query node modules (globally installed or within a directory).

```
select * from npm_packages;
select * from npm_packages where directory = "/home/andrew/test_project/";
```

Further work will need to be done for other operating systems.

obelisk commented on specs/linux/npm_packages.table:
```diff
@@ -0,0 +1,12 @@
+table_name("npm_packages")
+description("Lists all npm packages globally installed in a system.")
```
Or installed in a directory

obelisk commented on osquery/tables/system/linux/npm_packages.cpp:
```diff
@@ -0,0 +1,87 @@
+
+    Row r;
+    for (const auto& key_it : kPackageKeys) {
+      std::string key{key_it};
```
Do you need this copy here?

obelisk commented on specs/linux/npm_packages.table:
```diff
@@ -0,0 +1,12 @@
+    Column("description", TEXT, "Package supplied description"),
+    Column("author", TEXT, "Package author name"),
+    Column("license", TEXT, "License for package"),
+    Column("path", TEXT, "Path at to this module package.json"),
```
~at~ module's

osqueryer commented:
:-1: The commit e866f65e9ee23932630df51d49b20a66ef5f010f (Job results: [7011](https://jenkins.osquery.io/job/osqueryPullRequestLinux/7011/)) failed one or more tests (**Linux**).

osqueryer commented:
:-1: The commit e866f65e9ee23932630df51d49b20a66ef5f010f (Job results: [7012](https://jenkins.osquery.io/job/osqueryPullRequestLinux/7012/)) failed one or more tests (**Linux**).

osqueryer commented:
:-1: The commit 374a3ee899cbe52f59bd1a6e49d7c0b391930a04 (Job results: [3052](https://jenkins.osquery.io/job/osqueryPullRequestOSX/3052/)) failed one or more tests (**macOS/OS X**).

theopolis commented:
> obelisk added the do not merge  label 6 hours ago

Curious as to why?

obelisk commented:
@theopolis He is waiting to get the diff approved from his company

Exbia commented on osquery/tables/system/linux/npm_packages.cpp:
```diff
@@ -0,0 +1,86 @@
+    for (const auto& key : kPackageKeys) {
+      if (doc.doc().HasMember(key)) {
+        const auto& value = doc.doc()[key];
+        r[key] = (value.IsString()) ? value.GetString() : "";
```
Each key here should be a string. I think you're right, if a key is not the type expected then maybe we discard the credibility of the file and skip it with a log message.

osqueryer commented:
:-1: The commit 1c74a5e545c297eb42f245e66fc88c5f4ecd0571 (Job results: [3063](https://jenkins.osquery.io/job/osqueryPullRequestOSX/3063/)) failed one or more tests (**macOS/OS X**).

fmanco commented on osquery/tables/system/linux/npm_packages.cpp:
```diff
@@ -0,0 +1,110 @@
+
+    // Manually get nested key (Author name)
+    if (doc.doc().HasMember("author")) {
+      const auto& author = doc.doc()["author"]["name"];
```
Should you be checking the key "name" exists before accessing it?

Exbia replied:
Yes, I've fixed it now

fmanco commented on osquery/tables/system/linux/npm_packages.cpp:
```diff
@@ -0,0 +1,110 @@
+        }
+      }
+
+      const auto& author = doc.doc()["author"]["name"];
```
Is this supposed to be repeated here?

Exbia replied:
No it was not, fixed

fmanco commented on osquery/tables/system/linux/npm_packages.cpp:
```diff
@@ -0,0 +1,110 @@
+      } else {
+        // If its not a string, is it a dict with 'url' ?
+        if (license.HasMember("url")) {
+          const auto& license_url = doc.doc()["license"]["url"];
```
I didn't look up the interface but can you do `license["url"]`?

Exbia replied:
Yup, much more readable too, fixed
